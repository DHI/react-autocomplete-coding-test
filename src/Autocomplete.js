import React from 'react';

import debounce from 'lodash/debounce';

export const DEBOUNCE_DELAY = 500;

export default function Autocomplete() {
  return (
    <div className="Autocomplete root">
      <input className="input" placeholder="Start typing to search..." />
      <div className="dropdown"></div>
    </div>
  );
}
