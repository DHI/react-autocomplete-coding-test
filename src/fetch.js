import { shuffle } from 'lodash';

export const FETCH_DELAY = 1000;

const euCountries = [
  'Austria',
  'Belgium',
  'Bulgaria',
  'Croatia',
  'Cyprus',
  'Czechia',
  'Denmark',
  'Estonia',
  'Finland',
  'France',
  'Germany',
  'Greece',
  'Hungary',
  'Ireland',
  'Italy',
  'Kingdom',
  'Latvia',
  'Lithuania',
  'Luxembourg'
  // "Malta",
  // "Netherlands",
  // "Poland",
  // "Portugal",
  // "Romania",
  // "Slovakia",
  // "Slovenia",
  // "Spain",
  // "Sweden",
  // "United"
];

const euCountries2 = [
  // "Austria",
  // "Belgium",
  // "Bulgaria",
  // "Croatia",
  // "Cyprus",
  // "Czechia",
  // "Denmark",
  // "Estonia",
  // "Finland",
  // "France",
  'Germany',
  'Greece',
  'Hungary',
  'Ireland',
  'Italy',
  'Kingdom',
  'Latvia',
  'Lithuania',
  'Luxembourg',
  'Malta',
  'Netherlands',
  'Poland',
  'Portugal',
  'Romania',
  'Slovakia',
  'Slovenia',
  'Spain',
  'Sweden',
  'United'
];

global.fetch = url => {
  const [baseUrl, query] = url.split('?q=');

  if (!query) {
    return response(shuffle(euCountries));
  }

  const searchText = query.trim().toLowerCase();

  switch (baseUrl) {
    case 'https://countries.eu':
      if (searchText === 'p') {
        return Promise.reject('Oops! Something went wrong!');
      }
      return response(filter(searchText, euCountries));

    case 'https://countries2.eu':
      const data = filter(searchText, euCountries2);
      return response([...data, ...data]);

    default:
      return response(null, 'Invalid URL!');
  }
};

const response = (data, error) =>
  error
    ? Promise.reject(error)
    : new Promise(resolve =>
        setTimeout(() => {
          resolve({
            json: () => Promise.resolve(data)
          });
        }, FETCH_DELAY)
      );

const filter = (searchText, countries) =>
  Object.values(
    countries.reduce((acc, country) => {
      const countryLower = country.toLowerCase();
      if (countryLower.startsWith(searchText)) {
        acc[country] = {
          id: countryLower.substr(0, 3),
          name: country
        };
      }
      return acc;
    }, {})
  );
