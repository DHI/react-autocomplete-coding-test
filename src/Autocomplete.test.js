import React from 'react';

import { render, fireEvent, waitForElement, act } from '@testing-library/react';

import './fetch';
import { FETCH_DELAY } from './fetch';

import Autocomplete, { DEBOUNCE_DELAY } from './Autocomplete';

test('first fetch', async () => {
  const { container, getByPlaceholderText } = render(<Autocomplete />);

  const autocomplete = container.firstChild;
  const input = getByPlaceholderText(/start typing/i);

  await act(async () => {
    enterValue(input, 'a');
    await wait(100);
    enterValue(input, 'au');
    await wait(100);
    enterValue(input, 'aus');
  });

  expect(autocomplete).toMatchSnapshot('debounce should prevent any update');

  await act(async () => {
    await wait(600);
    enterValue(input, 'aust');
  });

  expect(autocomplete).toMatchSnapshot(
    'after debouching, the loader should be visible'
  );

  await act(() => waitDropdownTimeout());

  expect(autocomplete).toMatchSnapshot(
    'the first fetch should return the matched item'
  );
});

test('second fetch', async () => {
  const { container, getByPlaceholderText } = render(<Autocomplete />);

  const autocomplete = container.firstChild;
  const input = getByPlaceholderText(/start typing/i);

  act(() => {
    enterValue(input, 'xxx');
  });

  await waitDropdownOpen(input);

  expect(autocomplete).toMatchSnapshot('should display the not found message');
});

test('item not found', async () => {
  const { container, getByPlaceholderText } = render(<Autocomplete />);

  const autocomplete = container.firstChild;
  const input = getByPlaceholderText(/start typing/i);

  act(() => {
    enterValue(input, 'rom');
  });

  await waitDropdownOpen(input);

  expect(autocomplete).toMatchSnapshot();
});

test('from a value to an empty value', async () => {
  const { container, getByPlaceholderText } = render(<Autocomplete />);

  const autocomplete = container.firstChild;
  const input = getByPlaceholderText(/start typing/i);

  await act(async () => {
    enterValue(input, 'a');
    await wait(100);
    enterValue(input, '');
    waitDropdownTimeout();
  });

  expect(autocomplete).toMatchSnapshot('empty value should not update');
});

test('value delete', async () => {
  const { container, getByPlaceholderText } = render(<Autocomplete />);

  const autocomplete = container.firstChild;
  const input = getByPlaceholderText(/start typing/i);

  await act(async () => {
    enterValue(input, 'a');
    await waitDropdownOpen(input);
    enterValue(input, '');
    await wait(1550);
  });

  expect(autocomplete).toMatchSnapshot(
    'empty value, dropdown should not be open'
  );
});

test('first fetch fails', async () => {
  const fetchMock = jest.fn().mockImplementation(url => {
    if (/countries.eu/i.test(url)) {
      throw new Error('First fetch fails!');
    }

    return Promise.resolve({
      json: () =>
        Promise.resolve([
          {
            id: 'aus',
            name: 'Austria'
          }
        ])
    });
  });

  const origFetch = global.fetch;
  global.fetch = fetchMock;

  try {
    const { container, getByPlaceholderText } = render(<Autocomplete />);

    const autocomplete = container.firstChild;
    const input = getByPlaceholderText(/start typing/i);
    const queryValue = 'a';

    await act(async () => {
      enterValue(input, queryValue);
      await waitDebounceTimeout();
    });

    expectBothFetchesToBeCalled(fetchMock, queryValue);

    expect(autocomplete).toMatchSnapshot('should show the second fetch result');
  } finally {
    global.fetch = origFetch;
  }
});

test('first fetch fails in json call', async () => {
  const fetchMock = jest.fn().mockImplementation(url => {
    if (/countries.eu/i.test(url)) {
      return Promise.resolve({
        json: () => Promise.reject(new Error('First fetch fails!'))
      });
    }

    return Promise.resolve({
      json: () =>
        Promise.resolve([
          {
            id: 'aus',
            name: 'Austria'
          }
        ])
    });
  });

  const origFetch = global.fetch;
  global.fetch = fetchMock;

  try {
    const { container, getByPlaceholderText } = render(<Autocomplete />);

    const autocomplete = container.firstChild;
    const input = getByPlaceholderText(/start typing/i);
    const queryValue = 'a';

    await act(async () => {
      enterValue(input, queryValue);
      await waitDebounceTimeout();
    });

    expectBothFetchesToBeCalled(fetchMock, queryValue);

    expect(autocomplete).toMatchSnapshot('should show the second fetch result');
  } finally {
    global.fetch = origFetch;
  }
});

test('second fetch fails', async () => {
  console.error = jest.fn();

  const fetchMock = jest.fn().mockImplementation(url => {
    if (/countries2.eu/i.test(url)) {
      throw new Error('Second fetch fails');
    }

    return Promise.resolve({
      json: () =>
        Promise.resolve([
          {
            id: 'mal',
            name: 'Malta'
          }
        ])
    });
  });

  const origFetch = global.fetch;
  global.fetch = fetchMock;

  try {
    const { container, getByPlaceholderText } = render(<Autocomplete />);

    const autocomplete = container.firstChild;
    const input = getByPlaceholderText(/start typing/i);
    const queryValue = 'm';

    await act(async () => {
      enterValue(input, queryValue);
      await waitDebounceTimeout();
    });

    expectBothFetchesToBeCalled(fetchMock, queryValue);

    expect(console.error.mock.calls.length).toBe(0);

    expect(autocomplete).toMatchSnapshot('should show the first fetch result');
  } finally {
    global.fetch = origFetch;
  }
});

test('second fetch fails in json call', async () => {
  console.error = jest.fn();

  const fetchMock = jest.fn().mockImplementation(url => {
    if (/countries2.eu/i.test(url)) {
      return Promise.resolve({
        json: () => Promise.reject(new Error('Second fetch fails'))
      });
    }

    return Promise.resolve({
      json: () =>
        Promise.resolve([
          {
            id: 'mal',
            name: 'Malta'
          }
        ])
    });
  });

  const origFetch = global.fetch;
  global.fetch = fetchMock;

  try {
    const { container, getByPlaceholderText } = render(<Autocomplete />);

    const autocomplete = container.firstChild;
    const input = getByPlaceholderText(/start typing/i);
    const queryValue = 'm';

    await act(async () => {
      enterValue(input, queryValue);
      await waitDebounceTimeout();
    });

    expectBothFetchesToBeCalled(fetchMock, queryValue);

    expect(console.error.mock.calls.length).toBe(0);

    expect(autocomplete).toMatchSnapshot('should show the first fetch result');
  } finally {
    global.fetch = origFetch;
  }
});

test('first and second fetches fail', async () => {
  console.error = jest.fn();

  const fetchMock = jest.fn().mockImplementation(url => {
    throw new Error(
      (/countries.eu/i.test(url) ? 'First' : 'Second') + ' fetch fails'
    );
  });

  const origFetch = global.fetch;
  global.fetch = fetchMock;

  try {
    const { container, getByPlaceholderText } = render(<Autocomplete />);

    const autocomplete = container.firstChild;
    const input = getByPlaceholderText(/start typing/i);
    const queryValue = 'm';

    await act(async () => {
      enterValue(input, queryValue);
      await waitDebounceTimeout();
    });

    expectBothFetchesToBeCalled(fetchMock, queryValue);

    expect(console.error.mock.calls.length).toBe(1);
    expect(console.error.mock.calls[0][0].message).toBe(
      'Error fetching countries!'
    );

    expect(autocomplete).toMatchSnapshot('should not show any result');
  } finally {
    global.fetch = origFetch;
  }
});

test('item selection', async () => {
  const onSelectHandler = jest.fn();

  const { getByPlaceholderText, findByText } = render(
    <Autocomplete onSelect={onSelectHandler} />
  );

  const input = getByPlaceholderText(/start typing/i);

  act(() => {
    enterValue(input, 'a');
  });

  const itemFound = await findByText(/austria/i);

  await act(async () => {
    fireEvent.click(itemFound);
  });

  expect(onSelectHandler.mock.calls.length).toBe(1);
  expect(onSelectHandler.mock.calls[0][0]).toBe('Austria');
});

const wait = ms => new Promise(resolve => setTimeout(resolve, ms));

const waitDropdownTimeout = () => wait(FETCH_DELAY + DEBOUNCE_DELAY + 100);

const waitDebounceTimeout = () => wait(DEBOUNCE_DELAY + 100);

const waitDropdownOpen = input => waitForElement(() => input.nextSibling);

const enterValue = (input, value) =>
  fireEvent.change(input, { target: { value } });

const expectBothFetchesToBeCalled = (fetchMock, queryValue) => {
  expect(fetchMock.mock.calls.length).toBe(2);
  expect(fetchMock.mock.calls[0][0]).toBe(
    `https://countries2.eu?q=${queryValue}`,
    'fetch "countries2" without waiting the response'
  );
  expect(fetchMock.mock.calls[1][0]).toBe(
    `https://countries.eu?q=${queryValue}`,
    'fetch "countries" and wait the response'
  );
};
