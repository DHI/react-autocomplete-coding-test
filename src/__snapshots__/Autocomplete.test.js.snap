// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`first and second fetches fail: should not show any result 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="m"
  />
</div>
`;

exports[`first fetch fails in json call: should show the second fetch result 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="a"
  />
  <div
    class="dropdown"
  >
    <div
      class="item"
    >
      Austria
    </div>
  </div>
</div>
`;

exports[`first fetch fails: should show the second fetch result 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="a"
  />
  <div
    class="dropdown"
  >
    <div
      class="item"
    >
      Austria
    </div>
  </div>
</div>
`;

exports[`first fetch: after debouching, the loader should be visible 1`] = `
<div
  class="Autocomplete root loader"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="aust"
  />
</div>
`;

exports[`first fetch: debounce should prevent any update 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="aus"
  />
</div>
`;

exports[`first fetch: the first fetch should return the matched item 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="aust"
  />
  <div
    class="dropdown"
  >
    <div
      class="item"
    >
      Austria
    </div>
  </div>
</div>
`;

exports[`from a value to an empty value: empty value should not update 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value=""
  />
</div>
`;

exports[`item not found 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="rom"
  />
  <div
    class="dropdown"
  >
    <div
      class="item"
    >
      Romania
    </div>
  </div>
</div>
`;

exports[`second fetch fails in json call: should show the first fetch result 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="m"
  />
  <div
    class="dropdown"
  >
    <div
      class="item"
    >
      Malta
    </div>
  </div>
</div>
`;

exports[`second fetch fails: should show the first fetch result 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="m"
  />
  <div
    class="dropdown"
  >
    <div
      class="item"
    >
      Malta
    </div>
  </div>
</div>
`;

exports[`second fetch: should display the not found message 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value="xxx"
  />
  <div
    class="dropdown"
  >
    <div
      class="item notFound"
    >
      Country not found!
    </div>
  </div>
</div>
`;

exports[`value delete: empty value, dropdown should not be open 1`] = `
<div
  class="Autocomplete root"
>
  <input
    class="input"
    placeholder="Start typing to search..."
    value=""
  />
</div>
`;
